File: LeetCode_Marathons/LCM2/Hard/Maximum_Score_Words_Formed_by_Letters/a_bitmask.py
class Solution:
    def getMask(self, number: int, maskSize: int) -> str:
        mask = ""
        while number != 0:
            mask = str(number % 2) + mask
            number //= 2
        mask = "0" * (maskSize - len(mask)) + mask
        return mask


    def maxScoreWords(self, words: List[str], letters: List[str], score: List[int]) -> int:
        n = len(words)
        MaskCount = 2 ** n

        ans, scoreCounter = 0, 0
        mpLetters = defaultdict(int)

        for myChr in letters: mpLetters[myChr] += 1

        for intMask in range(0, MaskCount):
            mask = self.getMask(intMask, n)

            mpLettersOfMask = defaultdict(int)
            for indexOfMask in range(n):
                if mask[indexOfMask] == "1":
                    for myChr in words[indexOfMask]: mpLettersOfMask[myChr] += 1

            scoreCounter = 0
            for myChr in mpLettersOfMask.keys():
                if mpLettersOfMask[myChr] > mpLetters[myChr]:
                    scoreCounter = 0
                    break
                else: scoreCounter += score[ ord(myChr) - ord('a') ] * mpLettersOfMask[myChr]

            if scoreCounter > ans: ans = scoreCounter

        return ans

File: LeetCode_Marathons/LCM2/Hard/Maximum_Score_Words_Formed_by_Letters/a_bitmask.go
func getMask(number int, maskSize int) string {
    mask := ""
    for number != 0 {
        mask = strconv.Itoa(number % 2) + mask
        number /= 2
    }
    mask = strings.Repeat("0", maskSize - len(mask)) + mask
    return mask
}

func maxScoreWords(words []string, letters []byte, score []int) int {
    n := len(words)
    maskCount := int(math.Pow(float64(2), float64(n)))

    ans, scoreCounter := 0, 0
    mpLetters := make(map[byte]int)

    for _, myRune := range letters  {
        mpLetters[myRune]++
    }

    for i := 1; i < maskCount; i++ {
        mask := getMask(i, n)
        
        mpLettersOfMask := make(map[byte]int)
        for indexOfMask := 0; indexOfMask < n; indexOfMask++ {
            if mask[indexOfMask] == '1' {
                for _, myChr := range words[indexOfMask] {
                    mpLettersOfMask[byte(myChr)]++
                }
            }
        }

        scoreCounter = 0
        for key, value := range mpLettersOfMask {
            value2 := mpLetters[key]
            if value > value2 {
                scoreCounter = 0
                break
            } else {
                scoreCounter += score[ rune(key) - rune('a') ] * value
            }
        }
        
        if scoreCounter > ans {
            ans = scoreCounter
        }
    }

    return ans
}